src/Voron/Util/DataCopier.cs

```diff
@@ -36,51 +35,39 @@ public void ToStream(byte* ptr, long count, Stream output)
             }
         }
 
-        public void ToStream(AbstractPager src, long startPage, long numberOfPages, 
-            Stream output, Action<string> infoNotify, CancellationToken cancellationToken)
+        public void ToStream(Stream source, Stream output, long numberOfPages, Action<string> infoNotify, CancellationToken cancellationToken)
         {
-            // In case of encryption, we don't want to decrypt the data for backup, 
-            // so let's work directly with the underlying encrypted data (Inner pager).
+            // we're just copying the entire data file
          
             if ((_buffer.Length % Constants.Storage.PageSize) != 0)
                 throw new ArgumentException("The buffer length must be a multiple of the page size");
 
-            var steps = _buffer.Length/ Constants.Storage.PageSize;
             long totalCopied = 0;
             var toBeCopied = new Size(numberOfPages * Constants.Storage.PageSize, SizeUnit.Bytes).ToString();
             var totalSw = Stopwatch.StartNew();
             var sw = Stopwatch.StartNew();
 
-            using (var tempTx = new TempPagerTransaction())
-            fixed (byte* pBuffer = _buffer)
+            int copiedInBytes;
+            while ((copiedInBytes = source.Read(_buffer, 0, _buffer.Length)) != 0)
             {
-                for (var i = startPage; i < startPage + numberOfPages; i += steps)
-                {
-                    cancellationToken.ThrowIfCancellationRequested();
+                cancellationToken.ThrowIfCancellationRequested();
 
-                    var pagesToCopy = (int) (i + steps > numberOfPages ? numberOfPages - i : steps);
-                    src.EnsureMapped(tempTx, i, pagesToCopy);
-                    var ptr = src.AcquireRawPagePointer(tempTx, i);
```

Comment:
When we copy the data file we aren't under the write transaction lock. This means that we can have writes do data file meanwhile.

While our current implementation doesn't do anything special about that I'm not sure if that's correct. It'd require deeper investigation but I wonder if we shouldn't pass a pager state to `AcquireRawPagePointer()` which was taken via `GetPagerStateAndAddRefAtomically()` at the beginning of backup operation.  

It was implemented like that on the beginning: https://github.com/ravendb/ravendb/commit/316f2ffa2f13e25d18011f49e51136ed5182c181#diff-558ab70f893c6cff293b7c5f3d5cacf474b332b3caa93599e87c0b69155c4f2fL159

After thinking more about it I have some doubts if we shouldn't go this way. Note that the pager state might get changed during the backup if we allocate more pages.

Anyway, the new implementation is not gonna take that into account since it's not using mmap file. BTW How does it work if a file is enlarged meanwhile?

Reply:
> When we copy the data file we aren't under the write transaction lock.

We are. It's here:

https://github.com/ravendb/ravendb/blob/798e73374df099e4d52a096863d8a2e312e2079d/src/Voron/Impl/Backup/FullBackup.cs#L108